sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java

```diff
@@ -63,8 +87,87 @@ public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, P
+                List<Protos.Offer> offers = offerQueue.takeAll();
+                processOfferSet(offers);
+                synchronized (inProgressLock) {
+                    offersInProgress.removeAll(
+                            offers.stream()
+                                    .map(offer -> offer.getId())
+                                    .collect(Collectors.toList()));
+
+                    LOGGER.info("Offers in progress: {}", offersInProgress.stream().collect(Collectors.toList()));
+                }
+            }
+        });
+    }
+
+    protected abstract void processOfferSet(List<Protos.Offer> offers);
+
+    @Override
+    public void resourceOffers(SchedulerDriver driver, List<Protos.Offer> offers) {
+        if (!apiServerReady()) {
+            LOGGER.info("Waiting for API Server to start ...");
+            OfferUtils.declineOffers(driver, offers);
+            return;
+        }
+
+        // Task Reconciliation:
+        // Task Reconciliation must complete before any Tasks may be launched.  It ensures that a Scheduler and
+        // Mesos have agreed upon the state of all Tasks of interest to the scheduler.
+        // http://mesos.apache.org/documentation/latest/reconciliation/
+        reconciler.reconcile(driver);
+        if (!reconciler.isReconciled()) {
+            LOGGER.info("Waiting for task reconciliation to complete...");
+            OfferUtils.declineOffers(driver, offers);
+            return;
+        }
+
+        synchronized (inProgressLock) {
+            offersInProgress.addAll(
+                    offers.stream()
+                            .map(offer -> offer.getId())
+                            .collect(Collectors.toList()));
+        }
+
+        for (Protos.Offer offer : offers) {
+            boolean queued = offerQueue.offer(offer);
+            if (!queued) {
+                LOGGER.warn(
+                        "Failed to enqueue offer. Offer queue is full. Declining offer: '{}'",
+                        offer.getId().getValue());
+                OfferUtils.declineOffers(driver, Arrays.asList(offer));
+            }
+        }
+    }
+
     protected abstract void initialize(SchedulerDriver driver) throws InterruptedException;
 
+    protected abstract boolean apiServerReady();
+
+    /**
+     * All offers must have been presented to resourceOffers() before calling this.  This call will block until all
+     * offers have been processed.
+     * @throws InterruptedException
+     */
+    @VisibleForTesting
+    public void awaitOffersProcessed() throws InterruptedException {
+        while (true) {
+            synchronized (offersInProgress) {
+                if (offersInProgress.isEmpty()) {
+                    LOGGER.info("All offers processed.");
+                    return;
+                }
+            }
+
+            LOGGER.warn("Offers to be processed {} is non empty, sleeping for 500ms ...", offersInProgress);
```

Comment:
I mean, the logger message where it's printing offersInProgress should probably be within the lock protecting that. Feel free to defer to separate PR